Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,313 +41,590 @@ public sealed class RCDSystem : EntitySystem
-            gridId = location.GetGridUid(EntityManager);
-            // Check if fixing it failed / get final grid ID
-            if (!HasComp<MapGridComponent>(gridId))
-                return;
+            // Deconstructing an object
+            if (args.Target != null)
+            {
+                if (TryComp<RCDDeconstructableComponent>(args.Target, out var destructible))
+                {
+                    cost = destructible.Cost;
+                    delay = destructible.Delay;
+                    effectPrototype = destructible.Effect;
+                }
+            }
+
+            // Deconstructing a tile
+            else
+            {
+                var tile = _mapSystem.GetTileRef(mapGridData.Value.GridUid, mapGridData.Value.Component, mapGridData.Value.Location);
+                var protoName = (tile.GetContentTileDefinition().ID != "Lattice") ? "DeconstructTile" : "DeconstructLattice";
+
+                if (_protoManager.TryIndex<RCDPrototype>(protoName, out var deconProto))
+                {
+                    cost = deconProto.Cost;
+                    delay = deconProto.Delay;
+                    effectPrototype = deconProto.Effect;
+                }
+            }
+        }
+
+        // Tile construction modifiers
+        else if (component.CachedPrototype.Mode == RcdMode.ConstructTile)
+        {
+            // If replacing a tile, make the construction instant
+            var tile = _mapSystem.GetTileRef(mapGridData.Value.GridUid, mapGridData.Value.Component, mapGridData.Value.Location);
+
+            if (!tile.Tile.IsEmpty)
+            {
+                delay = 0;
+                effectPrototype = "EffectRCDConstruct0";
+            }
         }
 
-        var doAfterArgs = new DoAfterArgs(EntityManager, user, comp.Delay, new RCDDoAfterEvent(GetNetCoordinates(location), comp.Mode), uid, target: args.Target, used: uid)
+        #endregion
+
+        // Try to start the do after
+        var effect = Spawn(effectPrototype, mapGridData.Value.Location);
+        var ev = new RCDDoAfterEvent(GetNetCoordinates(mapGridData.Value.Location), component.ProtoId, cost, EntityManager.GetNetEntity(effect));
+
+        var doAfterArgs = new DoAfterArgs(EntityManager, user, delay, ev, uid, target: args.Target, used: uid)
         {
             BreakOnDamage = true,
-            NeedHand = true,
             BreakOnHandChange = true,
             BreakOnUserMove = true,
             BreakOnTargetMove = args.Target != null,
-            AttemptFrequency = AttemptFrequency.EveryTick
+            AttemptFrequency = AttemptFrequency.EveryTick,
+            CancelDuplicate = false,
+            BlockDuplicate = false
         };
 
         args.Handled = true;
 
-        if (_doAfter.TryStartDoAfter(doAfterArgs) && _gameTiming.IsFirstTimePredicted)
-            Spawn("EffectRCDConstruction", location);
+        if (!_doAfter.TryStartDoAfter(doAfterArgs) && _net.IsServer)
```

Comment:
you already have an IsServer guard clause earlier. just as an aside, i feel like most of this is fine to run on the client. idk how much it matters.